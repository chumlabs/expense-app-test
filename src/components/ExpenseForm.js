// IMPORTS
// - libraries
import React, { Component } from 'react';
import moment from 'moment';
import 'react-dates/initialize'; // needed for react-dates
import { SingleDatePicker } from 'react-dates';
import { Redirect } from 'react-router-dom';

export default class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toFixed(2).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = date => {
    // note: date is a moment object
    // if (date) {
    //   this.setState(() => ({ createdAt: date }));
    // }
    this.setState(() => ({ createdAt: date }));
    // console.log(`change to date: `, date);
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
    // optional validation
    // if (!this.state.createdAt) {
    //   this.setState(() => ({ error: 'Please enter a valid date' }));
    // } else {
    //   this.setState(() => ({ error: '' }));
    // }
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description) {
      this.setState(() => ({ error: 'Please enter a description' }));
    } else if (!this.state.amount) {
      this.setState(() => ({ error: 'Please enter an amount' }));
    } else if (!this.state.createdAt) {
      this.setState(() => ({ error: 'Please enter a valid date' }));
    } else {
      this.setState(() => ({ error: '' }));
      // console.log('submitted!');
      // use onSubmit prop
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description (required)"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus
            required
          />
          <br />
          <input
            type="text"
            placeholder="Amount (required)"
            value={this.state.amount}
            onChange={this.onAmountChange}
            required
          />
          <br />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <br />
          <textarea
            name="notes"
            id=""
            cols="40"
            rows="3"
            placeholder="Additional Notes (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <br />
          <button>{this.props.actionText}</button>
          {this.state.error ? <p className="error-text">{this.state.error}</p> : null}
        </form>
      </div>
    );
  }
}

// <button onClick={() => this.props.history.push('/')}>Cancel</button>
// <input type="date" value={this.state.amount} onChange={this.onDateChange} />

// course version of code
// import React from 'react';
// import moment from 'moment';
// import 'react-dates/initialize';
// import { SingleDatePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';

// export default class ExpenseForm extends React.Component {
//   constructor(props) {
//     super(props);

//     this.state = {
//       description: props.expense ? props.expense.description : '',
//       note: props.expense ? props.expense.note : '',
//       amount: props.expense ? (props.expense.amount / 100).toString() : '',
//       createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
//       calendarFocused: false,
//       error: ''
//     };
//   }
//   onDescriptionChange = e => {
//     const description = e.target.value;
//     this.setState(() => ({ description }));
//   };
//   onNoteChange = e => {
//     const note = e.target.value;
//     this.setState(() => ({ note }));
//   };
//   onAmountChange = e => {
//     const amount = e.target.value;

//     if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
//       this.setState(() => ({ amount }));
//     }
//   };
//   onDateChange = createdAt => {
//     if (createdAt) {
//       this.setState(() => ({ createdAt }));
//     }
//   };
//   onFocusChange = ({ focused }) => {
//     this.setState(() => ({ calendarFocused: focused }));
//   };
//   onSubmit = e => {
//     e.preventDefault();

//     if (!this.state.description || !this.state.amount) {
//       this.setState(() => ({ error: 'Please provide description and amount.' }));
//     } else {
//       this.setState(() => ({ error: '' }));
//       this.props.onSubmit({
//         description: this.state.description,
//         amount: parseFloat(this.state.amount, 10) * 100,
//         createdAt: this.state.createdAt.valueOf(),
//         note: this.state.note
//       });
//     }
//   };
//   render() {
//     return (
//       <div>
//         {this.state.error && <p>{this.state.error}</p>}
//         <form onSubmit={this.onSubmit}>
//           <input
//             type="text"
//             placeholder="Description"
//             autoFocus
//             value={this.state.description}
//             onChange={this.onDescriptionChange}
//           />
//           <input
//             type="text"
//             placeholder="Amount"
//             value={this.state.amount}
//             onChange={this.onAmountChange}
//           />
//           <SingleDatePicker
//             date={this.state.createdAt}
//             onDateChange={this.onDateChange}
//             focused={this.state.calendarFocused}
//             onFocusChange={this.onFocusChange}
//             numberOfMonths={1}
//             isOutsideRange={() => false}
//           />
//           <textarea
//             placeholder="Add a note for your expense (optional)"
//             value={this.state.note}
//             onChange={this.onNoteChange}
//           />
//           <button>Add Expense</button>
//         </form>
//       </div>
//     );
//   }
// }
