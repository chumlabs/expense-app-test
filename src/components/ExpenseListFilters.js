// IMPORTS
// - libraries
import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';

import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
// - components
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filterActions';

export class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = e => {
    // console.log(e.target.value);
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
        <h3>Applied Filters</h3>
        Text search:
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
        Sort expenses by:
        <select
          name="sort"
          id="sort"
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">date</option>
          <option value="amount">amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
          startDateId={'start'}
          endDateId={'end'}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: date => dispatch(setStartDate(date)),
  setEndDate: date => dispatch(setEndDate(date)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
