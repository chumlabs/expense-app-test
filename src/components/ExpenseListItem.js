// IMPORTS
// - libraries
import React from 'react';
// import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
// - components
// import { removeExpense } from '../actions/expenseActions';

const ExpenseListItem = ({ id, description, note, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`} className="list-item">
      <div>
        <h3 className="list-item__title">{description} </h3>
        <span className="list-item__subtitle">
          {moment(createdAt).format('MMM Do, YYYY')}
        </span>
      </div>
      <div>
        <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
      </div>
    </Link>
  </div>
);

export default ExpenseListItem;
