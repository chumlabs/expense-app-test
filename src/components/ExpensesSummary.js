import React, { Component } from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotalAmount }) => {
  const expenseTense = expensesCount === 1 ? 'expense' : 'expenses';
  const formattedAmount = numeral(expensesTotalAmount / 100).format('$0,0.00');

  return (
    <div>
      <p>{`Showing ${expensesCount} ${expenseTense} totalling ${formattedAmount}`}</p>
    </div>
  );
};

const mapStateToProps = state => {
  const filteredExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expensesCount: filteredExpenses.length,
    expensesTotalAmount: selectExpensesTotal(filteredExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
