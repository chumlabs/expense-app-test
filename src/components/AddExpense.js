// IMPORTS
// - libraries
import React, { Component } from 'react';
import { connect } from 'react-redux';
// - components
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenseActions';

export class AddExpense extends Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} actionText="Add Expense" />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(null, mapDispatchToProps)(AddExpense);
