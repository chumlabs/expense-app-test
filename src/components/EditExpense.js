// IMPORTS
// - libraries
import React, { Component } from 'react';
import { connect } from 'react-redux';
// - components
import ExpenseForm from './ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenseActions';

export class EditExpense extends Component {
  // { dispatch, expense, match, history }
  onSubmit = exp => {
    // console.log('expense updated', exp);
    this.props.editExpense(this.props.expense.id, exp);
    this.props.history.push('/');
  };

  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <h2>Expensify - Edit Entry</h2>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
          actionText="Update Expense"
        />
        <button onClick={this.onClick}>Remove Expense</button>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  expense: state.expenses.find(exp => exp.id === ownProps.match.params.id)
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  editExpense: (id, exp) => dispatch(editExpense(id, exp)),
  startRemoveExpense: idObj => dispatch(startRemoveExpense(idObj))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
