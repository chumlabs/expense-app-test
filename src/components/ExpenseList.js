// IMPORTS
// -libraries
import React from 'react';
import { connect } from 'react-redux';
// - components
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
// import { removeExpense } from '../actions/expenseActions';

// presentational component
export const ExpenseList = ({ expenses }) => (
  <div>
    <h1>Expense List</h1>
    <p>You have {expenses ? expenses.length : 0} expenses</p>
    {expenses ? (
      expenses.map(exp => <ExpenseListItem {...exp} key={exp.id} />)
    ) : (
      <p>Create an expense or clear the date range filter to see additional expenses</p>
    )}
  </div>
);

// mapStateToProps func (convention)
const mapStateToProps = state => ({
  expenses: selectExpenses(state.expenses, state.filters)
});

// connect funct (HOC component)
// - composes presentational component with state
export default connect(mapStateToProps)(ExpenseList);

// --------------------------
// // alt version - attempting to used mapDispatchToProps to pass onClick as prop
// const ExpenseList = ({ expenses, onListItemClick }) => (
//   <div>
//     <h1>Expense List</h1>
//     {expenses.map(exp => (
//       <ExpenseListItem {...exp} key={exp.id} onClick={onListItemClick} />
//     ))}
//     {console.log(onListItemClick)}
//   </div>
// );

// const mapStateToProps = state => ({
//   expenses: selectExpenses(state.expenses, state.filters)
// });

// const mapDispatchToProps = (dispatch, ownProps) => ({
//   onListItemClick: (dispatch) => dispatch(removeExpense(ownProps.id))
// });

// export default connect(mapStateToProps, mapDispatchToProps)(ExpenseList);
