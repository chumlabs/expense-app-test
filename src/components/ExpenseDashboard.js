// IMPORTS
// - libraries
import React, { Component } from 'react';
// - components
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboard = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
