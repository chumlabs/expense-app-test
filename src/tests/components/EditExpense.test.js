import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense } from '../../components/EditExpense';
import expenses from '../fixtures/expenseData';

// test setup
let expense, editExpense, history, removeExpense, wrapper;
beforeEach(() => {
  expense = expenses[0];
  editExpense = jest.fn();
  history = { push: jest.fn() };
  removeExpense = jest.fn();
  wrapper = shallow(
    <EditExpense
      editExpense={editExpense}
      history={history}
      removeExpense={removeExpense}
      expense={expense}
    />
  );
});

test('should render EditExpense component properly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit handler from ExpenseForm', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense);
});

test('should handle onClick handler', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith({ id: expense.id });
});
