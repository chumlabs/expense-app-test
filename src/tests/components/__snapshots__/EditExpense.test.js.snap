// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpense component properly 1`] = `
<div>
  <h2>
    Expensify - Edit Entry
  </h2>
  <ExpenseForm
    actionText="Update Expense"
    expense={
      Object {
        "amount": 700,
        "createdAt": 1491062400000,
        "description": "item 1 - abc",
        "id": "1",
        "note": "lowest amount, middle date",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove Expense
  </button>
</div>
`;
